Dwolla.Client/HttpServices/DwollaHttpService.cs

```diff
@@ -0,0 +1,113 @@
+﻿using Dwolla.Client.HttpRequestServices;
+using Dwolla.Client.HttpServices.Architecture;
+using Dwolla.Client.Models;
+using Dwolla.Client.Models.Responses;
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Dwolla.Client.HttpServices
+{
+	public class DwollaHttpService
+	{
+		private IDwollaClient _client;
+		public IDwollaClient Client
+		{
+			get => _client ?? DwollaConfiguration.Client;
+			set => _client = value;
+		}		
+		
+		private static TokenResponse _cachedToken;
+
+		private Task<string> getAccessTokenAsync;	
+
+		private async Task<string> GetAccessTokenAsync()
+		{
+			if (_cachedToken != null)//_cachedToken.ExpiresIn > now)
+			{ 
+				return _cachedToken.Token;
+			}
+
+			var response = await Authorization.GetToken();
+
+			if (response.Error != null)
+			{
+				throw new Exception(response.Error.Message);
+			}
+
+			_cachedToken = response.Content;
+
+			return _cachedToken.Token;
+		}
+
+		public DwollaHttpService()
+		{
+
+		}
+
+		public DwollaHttpService(Task<string> getAccessTokenAsync)
+		{
+			this.getAccessTokenAsync = getAccessTokenAsync;
+		}
+
+		public DwollaHttpService(IDwollaClient dwollaClient)
+		{
+			_client = dwollaClient;
+		}
+
+		public DwollaHttpService(IDwollaClient dwollaClient, Task<string> getAccessTokenAsync)
+		{
+			this.getAccessTokenAsync = getAccessTokenAsync;
+			_client = dwollaClient;
+		}
+
+		public async Task CacheAccessTokenAsync()
+		{
+			await GetAccessTokenAsync();
+		}
+
+		private AuthorizationHttpService _authorization;
+		public AuthorizationHttpService Authorization
+		{
+			get 
+			{
+				if (_authorization is null)
+				{
+					_authorization = new AuthorizationHttpService(Client, getAccessTokenAsync);
```

Comment:
I think you'll notice when refactoring the `Action<>` that this `getAccessTokenAsync` is different from what is passed to the other ones